src/core/thread/mle_router.cpp

```diff
@@ -4482,6 +4483,66 @@ Error MleRouter::SendTimeSync(void)
 }
 #endif // OPENTHREAD_CONFIG_TIME_SYNC_ENABLE
 
+Error MleRouter::DetachGracefully(otDetachGracefullyCallback aCallback, void *aContext)
+{
+    Error error = kErrorNone;
+
+    if (IsChild())
+    {
+        ExitNow(error = Mle::DetachGracefully(aCallback, aContext));
+    }
+
+    OT_ASSERT(IsDisabled() || IsDetached() || IsRouter() || IsLeader());
+
+    if (mDetachGracefullyTimer.IsRunning())
+    {
+        ExitNow(error = kErrorBusy);
+    }
+
+    OT_ASSERT(mDetachGracefullyCallback == nullptr);
+
+    mDetachGracefullyCallback = aCallback;
+    mDetachGracefullyContext  = aContext;
+
+    mDetachGracefullyTimer.Start(kDetachGracefullyTimeout);
+
+    if (IsRouter())
+    {
+        SendAddressRelease(&MleRouter::HandleDetachGracefullyAddressReleaseResponse, this);
+        ExitNow(error = kErrorNone);
+    }
+    else
+    {
+        mStopThreadTask.Post();
+        ExitNow(error = kErrorNone);
+    }
+
+exit:
+    return error;
+}
+
+void MleRouter::HandleDetachGracefullyAddressReleaseResponse(void *               aContext,
+                                                             otMessage *          aMessage,
+                                                             const otMessageInfo *aMessageInfo,
+                                                             Error                aResult)
+{
+    static_cast<MleRouter *>(aContext)->HandleDetachGracefullyAddressReleaseResponse(aMessage, aMessageInfo, aResult);
+}
+
+void MleRouter::HandleDetachGracefullyAddressReleaseResponse(otMessage *          aMessage,
+                                                             const otMessageInfo *aMessageInfo,
+                                                             Error                aResult)
+{
+    OT_UNUSED_VARIABLE(aMessage);
+    OT_UNUSED_VARIABLE(aMessageInfo);
+    OT_UNUSED_VARIABLE(aResult);
+
+    if (mDetachGracefullyTimer.IsRunning())
```

Comment:
```suggestion
    if (IsDetachingGracefully())
```
nit: may use a method to encapsulate the logic (more readable).

Reply:
Done.